src/gsy_e/models/market/forward.py

```diff
@@ -0,0 +1,138 @@
+
+    @staticmethod
+    @abstractmethod
+    def _get_start_time(current_time: DateTime) -> DateTime:
+        """Return time when the market block starts."""
+
+    @staticmethod
+    @abstractmethod
+    def _get_end_time(current_time: DateTime) -> DateTime:
+        """Return time when the market block ends."""
+
+    @staticmethod
+    @abstractmethod
+    def _get_market_slot_duration(current_time: DateTime, _config) -> duration:
+        """Return duration of market slots inside the market block."""
+
+    def create_future_market_slots(self, current_market_time_slot: DateTime,
+                                   config: "SimulationConfig") -> None:
+        if not ConstSettings.ForwardMarketSettings.ENABLE_FORWARD_MARKETS:
+            return
+        self._create_future_market_slots(self._get_start_time(current_market_time_slot),
+                                         self._get_end_time(current_market_time_slot), config)
+
+
+class DayForwardMarket(ForwardMarketBase):
+    """Day forward market implementation"""
+
+    @staticmethod
+    def _get_start_time(current_time: DateTime) -> DateTime:
+        return current_time.set(hour=0, minute=0).add(days=1)
+
+    @staticmethod
+    def _get_end_time(current_time: DateTime) -> DateTime:
+        return current_time.set(hour=0, minute=0).add(weeks=1, days=1, hours=-1)
+
+    @staticmethod
+    def _get_market_slot_duration(_current_time: DateTime, _config) -> duration:
+        return duration(hours=1)
+
+
+class WeekForwardMarket(ForwardMarketBase):
+    """Week forward market implementation"""
+
+    @staticmethod
+    def _get_start_time(current_time: DateTime) -> DateTime:
+        days_until_next_monday = 7 - (current_time.day_of_week - 1)
+        return current_time.set(hour=0, minute=0).add(days=days_until_next_monday)
+
+    @staticmethod
+    def _get_end_time(current_time: DateTime) -> DateTime:
+        return current_time.set(hour=0, minute=0).add(years=1)
+
+    @staticmethod
+    def _get_market_slot_duration(_current_time: DateTime, _config) -> duration:
+        return duration(weeks=1)
+
+
+class MonthForwardMarket(ForwardMarketBase):
+    """Month forward market implementation"""
+
+    @staticmethod
+    def _get_start_time(current_time: DateTime) -> DateTime:
+        return current_time.set(day=1, hour=0, minute=0).add(months=1)
+
+    @staticmethod
+    def _get_end_time(current_time: DateTime) -> DateTime:
+        return current_time.set(day=1, hour=0, minute=0).add(years=2)
+
+    @staticmethod
+    def _get_market_slot_duration(current_time: DateTime, _config) -> duration:
+        number_of_days_month = monthrange(current_time.year, current_time.month)[1]
+        return duration(days=number_of_days_month)
+
+
+class YearForwardMarket(ForwardMarketBase):
+    """Year forward market implementation"""
+
+    @staticmethod
+    def _get_start_time(current_time: DateTime) -> DateTime:
+        return current_time.set(month=1, day=1, hour=0, minute=0).add(years=2)
```

Comment:
@spyrostz can we have your feedback on this?
If today is `2021-01-01 00:00`, when should the "year-forward" market start? From what I read in the docs, I'd say it should start on `2027-01-01 00:00` (1 year gap + 5 years duration). Is this correct?

Reply:
> Based on the description of the task, the yearly market should start 1 year in advance and should last for 5 years. With these in mind, Given the current time is: 2021-01-01 00:00, the _get_start_time function will return 2023-01-01 00:00 which is more than 1 years ahead.

This is the correct implementation as far as I see. The yearly market does not operate on a rolling-window fashion, but on calendar years. Therefore the start time of a market, no matter when it opens, should be at 01-01-202*. As for why there is no open market for 01-01-2022, I am copying from the requirements document:

`For a yearly product, the market opens 5 years in advance and closes 1 year before 1st day of the year of T0. This means that if you are to sell a product now, you can only sell for 2024-2028 because the yearly period starts on the 1st of the calendar year.`

Therefore I think that there is no problem with this part of the code and it adheres to the requirements.

Reply:
@spyrostz , I got your point but what the code returns doesn't match the years you mentioned:
> Hence, the market responsible for the 2026 energy product remains open for 2021, 2022, 2023 and 2024
![Screenshot from 2022-07-20 11-31-20](https://user-images.githubusercontent.com/31365000/179949265-cc353053-6cfe-4a45-9664-66f1dd85f2c4.png)